### PR TITLE
fix: error return decoration mode

### PR DESCRIPTION
**Pull request by Groveer**

### CODE REVIEW

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 1]
```diff
@@ -63,8 +63,20 @@ void WQuickXdgDecorationManagerPrivate::updateDecorationMode(QWXdgToplevelDecora
...
+            mode = WQuickXdgDecorationManager::None;
+            break;
+        case WLR_XDG_TOPLEVEL_DECORATION_V1_MODE_CLIENT_SIDE:
+            mode = WQuickXdgDecorationManager::Client;
+            break;
+        case WLR_XDG_TOPLEVEL_DECORATION_V1_MODE_SERVER_SIDE:
+            mode = WQuickXdgDecorationManager::Server;
+            break;
+        default:
+            break;
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:78]:
```suggestion
            Q_UNREACHABLE();;
```

File: src/server/qtquick/private/wquickxdgdecorationmanager_p.h [hunk 2]
```diff
@@ -44,9 +44,10 @@ class WAYLIB_SERVER_EXPORT WQuickXdgDecorationManager : public WQuickWaylandServ
     explicit WQuickXdgDecorationManager(QObject *parent = nullptr);
 
     enum DecorationMode {
-        DecidesByClient,
-        PreferClientSide,
-        PreferServerSide,
+        Undefined,
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgdecorationmanager_p.h:47]: Add a commit to explain Undefined/None

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 3]
```diff
@@ -163,7 +175,13 @@ WQuickXdgDecorationManagerAttached::WQuickXdgDecorationManagerAttached(WSurface
 }
 
 bool WQuickXdgDecorationManagerAttached::serverDecorationEnabled() const {
-    return m_manager->modeBySurface(m_target) == WQuickXdgDecorationManager::PreferServerSide;
+    switch (m_manager->modeBySurface(m_target)) {
+    case WQuickXdgDecorationManager::None:
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:179]: The behavior of `None` should be customizable by the window manager

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp (hunk 3 above)
Comment (Groveer) [reply src/server/qtquick/private/wquickxdgdecorationmanager.cpp:179]: No, qt app send None, but it do't draw its self。

Comment (wineee) [reply src/server/qtquick/private/wquickxdgdecorationmanager.cpp:179]: Yon need call `decorat->setMode(WLR_XDG_TOPLEVEL_DECORATION_V1_MODE_CLIENT_SIDE);` after updateDecorationMode

File: src/server/qtquick/private/wquickxdgdecorationmanager_p.h [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
 #pragma once
 
+#include <qtmetamacros.h>
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgdecorationmanager_p.h:6]:
```suggestion
```

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickxdgdecorationmanager_p.h [hunk 5]
```diff
@@ -37,19 +38,24 @@ class WAYLIB_SERVER_EXPORT WQuickXdgDecorationManager : public WQuickWaylandServ
     W_DECLARE_PRIVATE(WQuickXdgDecorationManager)
     Q_PROPERTY(DecorationMode mode READ mode WRITE setMode NOTIFY modeChanged)
     Q_PROPERTY(WSocket* ownsSocket READ ownsSocket CONSTANT)
+    Q_PROPERTY(DecorationMode preferMode READ preferMode WRITE setPreferMode NOTIFY preferModeChanged)
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgdecorationmanager_p.h:41]: rename `preferMode` -> `preferredMode`

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 6]
```diff
@@ -92,6 +121,26 @@ void WQuickXdgDecorationManager::create()
...
+    return d->preferredMode;
+}
+
+void WQuickXdgDecorationManager::setPreferredMode(DecorationMode mode)
+{
+    W_D(WQuickXdgDecorationManager);
+    if (d->preferredMode == mode) {
+        return;
+    }
+    if (d->preferredMode < 1) {
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:136]: Should `< 2`?

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp (hunk 6 above)
Comment (Groveer) [reply src/server/qtquick/private/wquickxdgdecorationmanager.cpp:136]: Yes, should < 2

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 7]
```diff
@@ -34,13 +34,14 @@ class WQuickXdgDecorationManagerPrivate : public WObjectPrivate
...
     WQuickXdgDecorationManager::DecorationMode modeBySurface(WSurface *surface) const {
-        return decorations.value(surface, WQuickXdgDecorationManager::DecidesByClient);
+        return decorations.value(surface, WQuickXdgDecorationManager::Undefined);
     }
 
     W_DECLARE_PUBLIC(WQuickXdgDecorationManager)
 
     QWXdgDecorationManagerV1 *manager = nullptr;
-    WQuickXdgDecorationManager::DecorationMode mode = WQuickXdgDecorationManager::DecidesByClient;
+    WQuickXdgDecorationManager::DecorationMode mode = WQuickXdgDecorationManager::None;
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:43]: only need preferredMode

**Review by wineee — CHANGES_REQUESTED** [review]

**Review by wineee — APPROVED** [review]

**Review by zccrs — APPROVED** [review]

**Pull request merged**